db/stats.py

```diff
@@ -1,119 +1,244 @@
+        cur = connection.execute(q)
+        row = cur.fetchone()
+        if row:
+            return row[0]
+
+def get_next_hour(date):
+    """Round up a date to the nearest hour:00:00.
+    Arguments:
+      date: a datetime
+    """
+    delta = datetime.timedelta(hours=1)
+    date = date + delta
+    date = date.replace(minute=0, second=0, microsecond=0)
+    return date
+
+def compute_stats(to_date):
+    # Compute stats up to the given date
+
+    stats_date = get_most_recent_stats_date()
+    if not stats_date:
+        # If there have been no statistics, we start from the
+        # earliest date in the lowlevel table
+        stats_date = get_earliest_submission_date()
+        if not stats_date:
+            # If there are no lowlevel submissions, we stop
+            return
+
+    next_date = get_next_hour(stats_date)
+    with db.engine.connect() as connection:
+        while next_date < to_date:
+            stats = _count_submissions_to_date(connection, next_date)
+            write_stats(connection, next_date, stats)
+            next_date = get_next_hour(next_date)
+
+def write_stats(connection, date, stats):
+    for name, value in six.iteritems(stats):
+        """Records a value with a given name and current timestamp."""
+        q = text("""
+            INSERT INTO statistics (collected, name, value)
+                 VALUES (:collected, :name, :value)""")
+        connection.execute(q, {"collected": date, "name": name, "value": value})
 
 def get_stats():
-    stats_keys = ["lowlevel-lossy", "lowlevel-lossy-unique", "lowlevel-lossless", "lowlevel-lossless-unique"]
-    # TODO: Port this to new implementation:
-    stats = db.cache._mc.get_multi(stats_keys, key_prefix="ac-num-")
-    last_collected = db.cache.get('last-collected')
-
-    # Recalculate everything together, always.
-    if sorted(stats_keys) != sorted(stats.keys()) or last_collected is None:
-        stats_parameters = dict([(a, 0) for a in stats_keys])
+    """Get submission statistics based on low-level data.
+
+    In addition, all stats are recorded in the database when this function is
+    called if an hour has passed since last record.
+
+    Returns:
+        Dictionary with values for each metric:
+        - lowlevel-lossy: total number of lossy submissions
+        - lowlevel-lossless: total number of lossless submissions
+        - lowlevel-lossy-unique: number of unique lossy submissions
+        - lowlevel-lossless-unique: number of unique lossless submissions
+    """
+
+    # Names are also used as cache keys with prefix defined below.
+    stats_names = [
+        LOWLEVEL_LOSSY,
+        LOWLEVEL_LOSSLESS,
+        LOWLEVEL_LOSSY_UNIQUE,
+        LOWLEVEL_LOSSLESS_UNIQUE,
+        LOWLEVEL_TOTAL,
+        LOWLEVEL_TOTAL_UNIQUE,
+    ]
+    cache_key_prefix = "ac-num-"
+    # TODO: Port this to new implementation (don't use protected parts):
+    stats = db.cache._mc.get_multi(stats_names, key_prefix=cache_key_prefix)
```

Comment:
I think this was not working before because of exactly that issue in caching wrapper that you found. So it's probably safe to use our custom interface now.